src/transform/src/projection_pushdown.rs

```diff
@@ -0,0 +1,401 @@
+//! repeated columns reduce the width of data in the dataflow.
+//!
+//! Some comments have been inherited from the `Demand` transform.
+
+use std::collections::{HashMap, HashSet};
+
+use expr::{Id, JoinInputMapper, MirRelationExpr, MirScalarExpr};
+
+use crate::TransformArgs;
+
+/// Pushes projections down through other operators.
+#[derive(Debug)]
+pub struct ProjectionPushdown;
+
+impl crate::Transform for ProjectionPushdown {
+    fn transform(
+        &self,
+        relation: &mut MirRelationExpr,
+        _: TransformArgs,
+    ) -> Result<(), crate::TransformError> {
+        self.action(
+            relation,
+            (0..relation.arity()).collect(),
+            &mut HashMap::new(),
+        );
+        Ok(())
+    }
+}
+
+impl ProjectionPushdown {
+    /// Pushes the `desired_projection` down through `relation`.
+    ///
+    /// This action transforms `relation` to a `MirRelationExpr` equivalent to
+    /// `relation.project(desired_projection)`.
+    ///
+    /// `desired_projection` is expected to consist of unique columns.
+    fn action(
+        &self,
+        relation: &mut MirRelationExpr,
+        mut desired_projection: Vec<usize>,
+        gets: &mut HashMap<Id, HashSet<usize>>,
+    ) {
+        // Try to push the desired projection down through `relation`.
+        // In the process `relation` is transformed to a ` MirRelationExpr`
+        // equivalent to `relation.project(actual_projection)`.
+        let actual_projection = match relation {
+            // Nothing can be pushed through leaf nodes.
+            MirRelationExpr::Constant { .. } => (0..relation.arity()).collect(),
+            MirRelationExpr::Get { id, .. } => {
+                gets.entry(*id)
+                    .or_insert_with(HashSet::new)
+                    .extend(desired_projection.clone());
+                (0..relation.arity()).collect()
+            }
+            MirRelationExpr::Let { id, value, body } => {
+                // Let harvests any requirements of get from its body,
+                // and pushes the sorted union of the requirements at its value.
+                let id = Id::Local(*id);
+                let prior = gets.insert(id, HashSet::new());
+                self.action(body, desired_projection.clone(), gets);
+                let desired_value_projection = gets.remove(&id).unwrap();
+                if let Some(prior) = prior {
+                    gets.insert(id, prior);
+                }
+                let desired_value_projection = get_sorted_vec(desired_value_projection);
+                body.visit_mut_pre(&mut |e| {
+                    // When we push the `desired_value_projection` at `value`,
+                    // the columns returned by `Get(id)` will change, so we need
+                    // to permute `Project`s around `Get(id)`.
+                    if let MirRelationExpr::Project { input, outputs } = e {
+                        if let MirRelationExpr::Get { id: inner_id, .. } = &**input {
```

Comment:
Correct, we do nothing. If there's a `Get` without a projection around it, then the `desired_value_projection` is all the columns from the `Get`, and there is no need to permute anything.

Reply:
I'm super confused then. Is there a contract that `Get` which does not use all columns will certainly have a `Project` immediately around it? I could see that perhaps being true, as a consequence of how all of the rest of the code is structured, but if that is the contract that's important to write down.

Reply:
A thing I think I'm missing, and maybe this explanation would help, is:
> Having potentially removed columns from `value`, we need to correct references in `body`. This is non-trivial, because as the columns no longer exist there is not a correct substitution to perform (except perhaps to install `Datum::Dummy` values in the now absent locations using a `Map` and then permutation). Instead, we rely on the property that we discard columns only as a function of `Project` statements immediately guarding `Get` expressions, and for a column to be discarded all `Get` expressions in `body` must be contained in a `Project` that does not contain that column. We perform surgery between those to operators to ensure that no references to absent columns escape.

Reply:
I'm not certain I see where we ensure that we only register demanded columns from `Get` that come from a `Project`, but I could believe that we do make that happen, but it isn't clear to me. Probably something about how the only way to transform a `Get` to something "equivalent to `relation.project(desired_projection)`" is to install a `Project`, unlike for other operators where it can happen by pushing the projection into or through the operator, without leaving a `Project` behind.

Reply:
> Probably something about how the only way to transform a Get to something "equivalent to relation.project(desired_projection)" is to install a Project, unlike for other operators where it can happen by pushing the projection into or through the operator, without leaving a Project behind

Correct for all `desired_projection`s that are not just `0..typ.arity()`.